flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/RestartingTest.java

```diff
@@ -0,0 +1,223 @@
+                new StateValidator<>("WaitingForResources");
+
+        public void setExpectCancelling(Consumer<ExecutingTest.CancellingArguments> asserter) {
+            cancellingStateValidator.expectInput(asserter);
+        }
+
+        public void setExpectWaitingForResources() {
+            waitingForResourcesStateValidator.expectInput((none) -> {});
+        }
+
+        @Override
+        public void goToCanceling(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler) {
+            cancellingStateValidator.validateInput(
+                    new ExecutingTest.CancellingArguments(
+                            executionGraph, executionGraphHandler, operatorCoordinatorHandler));
+            hadStateTransition = true;
+        }
+
+        @Override
+        public void goToWaitingForResources() {
+            waitingForResourcesStateValidator.validateInput(null);
+            hadStateTransition = true;
+        }
+
+        @Override
+        public void runIfState(State expectedState, Runnable action, Duration delay) {
+            if (!hadStateTransition) {
+                action.run();
+            }
+        }
+
+        @Override
+        public void close() throws Exception {
+            super.close();
+            cancellingStateValidator.close();
+            waitingForResourcesStateValidator.close();
+        }
+    }
+
+    private static class CancellableExecutionGraph extends ExecutionGraph {
+        private boolean cancelled = false;
+
+        CancellableExecutionGraph() throws IOException {
+            super(
+                    new JobInformation(
+                            new JobID(),
+                            "Test Job",
+                            new SerializedValue<>(new ExecutionConfig()),
+                            new Configuration(),
+                            Collections.emptyList(),
+                            Collections.emptyList()),
+                    TestingUtils.defaultExecutor(),
+                    TestingUtils.defaultExecutor(),
+                    AkkaUtils.getDefaultTimeout(),
+                    1,
+                    ExecutionGraph.class.getClassLoader(),
+                    VoidBlobWriter.getInstance(),
+                    PartitionReleaseStrategyFactoryLoader.loadPartitionReleaseStrategyFactory(
+                            new Configuration()),
+                    NettyShuffleMaster.INSTANCE,
+                    NoOpJobMasterPartitionTracker.INSTANCE,
+                    ScheduleMode.EAGER,
+                    NoOpExecutionDeploymentListener.get(),
+                    (execution, newState) -> {},
+                    0L);
+        }
+
+        @Override
+        public void cancel() {
+            cancelled = true;
+        }
+
+        public boolean isCancelled() {
+            return cancelled;
+        }
+    }
```

Comment:
As a follow up we might wanna extract an interface from the `ExecutionGraph` to make it better testable. Now with every `ExecutionGraph` subclass we need to touch another place whenever we change something on the `ExecutionGraph` constructor.

Reply:
I filed a ticket for this: https://issues.apache.org/jira/browse/FLINK-21347